src/memory/AmdFlash.cc

```diff
@@ -47,21 +47,28 @@ AmdFlash::AmdFlash(const std::string& name, const ValidatedChip& validatedChip,
 	: motherBoard(config.getMotherBoard())
 	, chip(validatedChip.chip)
 {
-	auto numSectors = chip.geometry.sectorCount;
-	assert(writeProtectSectors.size() <= numSectors);
+	assert(writeProtectSectors.size() <= chip.geometry.sectorCount);
+
+	sectors.reserve(chip.geometry.sectorCount);
+	for (size_t address = 0; const Region& region : chip.geometry.regions) {
+		for (size_t regionSector = 0; regionSector < region.count; ++regionSector, address += region.size) {
+			sectors.push_back({
+				.address = address,
+				.size = region.size,
+				.writeProtect = sectors.size() < writeProtectSectors.size() && writeProtectSectors[sectors.size()]
+			});
+		}
+	}
```

Comment:
While reviewing this code, my immediate concern was that this only partially initializes the `Sector` objects. But when I continued reading I did see that also `writeAddress` and `readAddress` is correctly initialized.

Would it be cleaner to merge this loop and the loop below (the one that initializes `writeAddress`)? I mean it is technically possible, and not difficult, but I wonder if it makes the code cleaner.

And then the next question is: is it possible, and cleaner, to also include initialization of `readAddress` in the same loop? Maybe not.

Reply:
I don’t think it necessarily makes it cleaner. Now the code sections have a single responsibility. The write/read address logic is just a bit too complex to all blend together into a single loop.

Additionally I have a patch queued up which will actually get rid of `writeAddress` and `readAddress`, instead using a single TrackedRAM to store (and persist) the Flash memory. That will simplify things, but it’s not ready for submission yet. Keeping it as it is now will make that diff a bit cleaner.